WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormViewModelTests.swift

```diff
@@ -170,14 +171,37 @@ final class ProductFormViewModelTests: XCTestCase {
         XCTAssertEqual(viewModel.productModel.product.variations, newProduct.variations)
         XCTAssertEqual(viewModel.productModel.product.attributes, newProduct.attributes)
     }
+
+    func test_update_variations_fires_replace_product_action() throws {
+        // Given
+        let product = Product()
+        let mockStores = MockStoresManager(sessionManager: SessionManager.testingInstance)
+        let viewModel = createViewModel(product: product, formType: .edit, stores: mockStores)
+
+        // When
+        let attributes = ProductAttribute(siteID: 0, attributeID: 0, name: "Color", position: 0, visible: true, variation: true, options: ["Green, Blue"])
+        let newProduct = product.copy(attributes: [attributes], variations: [1])
+        viewModel.updateProductVariations(from: newProduct)
+
+        // Then
+        let action = try XCTUnwrap(mockStores.receivedActions.first as? ProductAction)
+        switch action {
+        case .replaceProductLocally:
+            break // Expected
+        default:
+            XCTFail("Unexpected action. Got \(action). Expected \(ProductAction.replaceProductLocally(product: newProduct, onCompletion: {}))")
+        }
```

Comment:
Really nice failure message! I find this `switch` just a little confusing since the `default` is the failure case (although the comment helps clarify it). Another option could be something like this:

```swift
guard case .replaceProductLocally = action else {
        return XCTFail("Unexpected action. Got \(action). Expected \(ProductAction.replaceProductLocally(product: newProduct, onCompletion: {}))")
}
```

But it certainly works as-is!

Reply:
Yeah, I really didn't like that test because we are asserting for failure and not for success.
I have changed to assert for success! https://github.com/woocommerce/woocommerce-ios/pull/3787/commits/ef271fe478293865326d7fe8e3bdb74a72015256